### PR TITLE
[ENH] adding slurm as a system type

### PR DESCRIPTION
**Pull request by djarecka**
- adding commands for slurm 
- some refactoring of `generate_job_submit_template`

### CODE REVIEW

**Comment by djarecka** [comment]
@zhao-cy - this is the draft of the PR, once again my apology that it took so long.

Example of the output of the slurm functions:
- sacct
```
[19:18][18.18][15%]djarecka@openmind7:babs$ sacct -u djarecka -j 30721579
JobID           JobName  Partition    Account  AllocCPUS      State ExitCode 
------------ ---------- ---------- ---------- ---------- ---------- -------- 
30721579     fmr_sub-02     normal     gablab          2  COMPLETED      0:0 
30721579.ba+      batch                gablab          2  COMPLETED      0:0 
30721579.ex+     extern                gablab          2  COMPLETED      0:0 
```
- squeue
```
[19:24][19.72][17%]djarecka@openmind7:babs$ squeue -u djarecka -o "%.18i %.9P %.8j %.8u %.2t %T %.10M"
             JOBID PARTITION     NAME     USER ST STATE       TIME
          30723103    normal fmr_sub- djarecka  R RUNNING       0:23
          30723104    normal fmr_sub- djarecka  R RUNNING       0:23
          30723105    normal fmr_sub- djarecka  R RUNNING       0:23
```
- output of sbtach:
```
Submitted batch job 30723107
```

**Comment by zhao-cy** [comment]
Thank you so much Dorota!! I'll take a look and try it out tomorrow!

**Comment by zhao-cy** [comment]
@djarecka I've made some changes and pushed a commit to this PR, mainly to fix job submission part and Slurm directives - see [here](https://github.com/PennLINC/babs/pull/92/commits/664325134413954f461722b0338017e9a2e9ff98). I'll continue reviewing your changes.

**Review by zhao-cy — COMMENTED** [review]
@djarecka I haven't fully finished reviewing yet, but as it's working on the basic use cases, I am thinking about merging it to main now, and let's create separate branches for further changes to avoid conflicts in the same branch. 

I will make a new branch and change the function names and variables names related to function `_parsing_sacct_out()` (which should be `squeue`, instead of `sacct`).

File: babs/dict_cluster_systems.yaml [hunk 1]
```diff
@@ -10,4 +10,11 @@ sge:
...
   temporary_disk_space: "-l tmpfree=$VALUE"   # "-l tmpfree=200G" on cubic
   number_of_cpus: "-pe threaded $VALUE"   # "-pe threaded N" or a range: "-pe threaded N-M", N<M on cubic
-  hard_runtime_limit: "-l h_rt=$VALUE"    # "-l h_rt=24:00:00" on cubic
\ No newline at end of file
+  hard_runtime_limit: "-l h_rt=$VALUE"    # "-l h_rt=24:00:00" on cubic
+slurm:
+  interpreting_shell: ""
+  hard_memory_limit: "--mem=$VALUE"
+  soft_memory_limit: ""
+  temporary_disk_space: ""
```
[the remark this answers is not in the document]
Comment (zhao-cy) [reply babs/dict_cluster_systems.yaml]: I've changed this in commit [6643251](https://github.com/PennLINC/babs/pull/92/commits/664325134413954f461722b0338017e9a2e9ff98), so no worries.

File: babs/babs.py [hunk 2]
```diff
@@ -2825,6 +2830,12 @@ def generate_job_submit_template(self, yaml_path, input_ds, babs, system):
             env_flags = "-v DSLOCKFILE=" + babs.analysis_path + "/.SGE_datalad_lock"
             eo_args = "-e " + babs.analysis_path + "/logs " \
                 + "-o " + babs.analysis_path + "/logs"
+        elif system.type == "slurm":
+            submit_head = "sbatch"
+            env_flags = "--export=DSLOCKFILE=" + babs.analysis_path + f"/.SLURM_datalad_lock,CBICA_TMPDIR={self.config['job_compute_space']}"
+            job_nm = self.container_name[0:3] + "_" + "${sub_id}"
```
[the remark this answers is not in the document]
Comment (zhao-cy) [reply babs/babs.py]: I've fixed this in commit [6643251](https://github.com/PennLINC/babs/pull/92/commits/664325134413954f461722b0338017e9a2e9ff98) so no worries about this.

File: babs/utils.py [hunk 3]
```diff
@@ -1663,26 +1689,54 @@ def request_all_job_status():
...
-    proc_qstat.check_returncode()
-    msg = proc_qstat.stdout.decode('utf-8')
-    print(msg)
+    std = sacct_proc.stdout.decode('utf-8')
+
+    sacct_out_df = _parsing_sacct_out(std)
+    return sacct_out_df
+
+
+def _parsing_sacct_out(sacct_std):
```
[the remark this answers is not in the document]
Comment (zhao-cy) [reply babs/utils.py:1708]: I will handle this problem after I merge the current PR into main + create a new branch.

**Comment by zhao-cy** [comment]
In this PR: commands for Slurm have been added. Partial tests have been done, including those on multi-session toy dataset:
* `babs-check-setup --job-test`
* `babs-submit` one session
* `babs-status`

Further tests need to be done; code reviews for corresponding functions need to be done - 
* single-session dataset;
* `babs-submit` other options
* `babs-status` other options, including:
  * `--resubmit`
  * failed job auditing, using `--container-config-yaml-file` and `--job-account`

File: babs/utils.py [hunk 4]
```diff
@@ -1663,26 +1689,54 @@ def request_all_job_status():
...
+    header_l = sacct_std.splitlines()[0].split()
+    datarows = sacct_std.splitlines()[1:]
```
Comment (zhao-cy) [inline babs/utils.py:1710]: There are scenarios where no job is running, so there is no 2nd line (`[1:]`)

File: babs/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (djarecka) [reply babs/utils.py:1708]: yes, it was sacct, but later I changed to `squeue` and forgot to update the name

File: babs/babs.py [hunk 5]
```diff
@@ -2825,6 +2830,12 @@ def generate_job_submit_template(self, yaml_path, input_ds, babs, system):
             env_flags = "-v DSLOCKFILE=" + babs.analysis_path + "/.SGE_datalad_lock"
             eo_args = "-e " + babs.analysis_path + "/logs " \
                 + "-o " + babs.analysis_path + "/logs"
+        elif system.type == "slurm":
+            submit_head = "sbatch"
+            env_flags = "--export=DSLOCKFILE=" + babs.analysis_path + f"/.SLURM_datalad_lock,CBICA_TMPDIR={self.config['job_compute_space']}"
```
[the remark this answers is not in the document]
Comment (djarecka) [reply babs/babs.py]: i think I had some issue without it, I will double check...

File: babs/utils.py
[the remark this answers is not in the document]
Comment (djarecka) [reply babs/utils.py:1]: somehow I didn't need it

File: babs/utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (zhao-cy) [reply babs/utils.py:1708]: No worries! It has been fixed in the new branch/PR #96.

File: babs/utils.py [hunk 6]
```diff
@@ -1939,6 +1995,43 @@ def check_job_account(job_id_str, job_name, username_lowercase):
...
+    get information for a finished job in Slurm by calling `sacct`
+    """
+    proc_qacct = subprocess.run(
+        ["sacct", "-u", username_lowercase,
+         "-j", job_id_str],
+        stdout=subprocess.PIPE
+    )
+
+    proc_qacct.check_returncode()
+    msg_l = proc_qacct.stdout.decode('utf-8').split("\n")
```
Comment (zhao-cy) [inline babs/utils.py:2015]: These should be `sacct` instead of `qacct`. I'm changing them in the PR #96.

File: babs/utils.py [hunk 7]
```diff
@@ -1939,6 +1995,43 @@ def check_job_account(job_id_str, job_name, username_lowercase):
...
+    msg_head = msg_l[0].split()
+    if "State" not in msg_head or "JobID" not in msg_head or "JobName" not in msg_head:
+        return "sacct doesn't provide information about the job"
```
Comment (zhao-cy) [inline babs/utils.py:2018]: Even if `sacct` is empty for a job, printed message will still include the head of the table - it looks like this on my end: I'm using a non-existing job ID from my account:
```
% sacct -j 12345 -u $USER
JobID           JobName  Partition    Account  AllocCPUS      State ExitCode 
------------ ---------- ---------- ---------- ---------- ---------- --------
```
So I'm adding additional checks like if there is a real job row in PR #96.

File: babs/utils.py [hunk 8]
```diff
@@ -1939,6 +1995,43 @@ def check_job_account(job_id_str, job_name, username_lowercase):
...
+
+    st_ind = msg_head.index("State")
+    jobid_ind = msg_head.index("JobID")
+    jobnm_ind = msg_head.index("JobName")
+    job_saact = msg_l[2].split() # the 2nd row should have the main job
+
+    if job_saact[jobid_ind] != job_id_str or job_saact[jobnm_ind] != job_name:
+        return "sacct doesn't have the info for the specific job or the format is different"
```
Comment (zhao-cy) [inline babs/utils.py:2026]: The column of 'JobName' can be very narrow to hold the entire job name. I'm making the columns wider in PR #96.

**Pull request merged**